posts/nix-vs-docker-i.md

```diff
@@ -0,0 +1,260 @@
+underlying models are so different that it's not a fair one-to-one comparison.
+Still, for a growing number of use cases, Nix doesn't just replace Docker. It
+does a better job. Nix, after all, is a superset of Docker.
+
+## The Build Model
+
+Dockerfiles are imperative. You declare a sequence of commands: set a base
+image, copy some files, install packages, maybe expose a port, and then define
+an entry point. These steps are executed one after another, and each one forms a
+layer on top of the last. The final image is a snapshot of the resulting
+filesystem.
+
+This process is simple but brittle. A change in one line can invalidate the
+entire build cache. Even when using pinned base images and lockfiles, there's
+always a risk that some transient state---like a registry being temporarily
+down, or a package mirror being updated---will corrupt reproducibility.
+
+Take this example:
+
+```dockerfile
+FROM node:20-alpine
+WORKDIR /app
+COPY package*.json ./
+RUN npm install
+COPY . .
+CMD ["node", "index.js"]
+```
+
+If `npm install` happens at build time, the image depends on the current state
+of the npm registry. A patch update upstream can change your build without
+notice. It's not deterministic, and it's not safe.
+
+Here's how the same thing works in Nix using a shell environment:
+
+```nix
+pkgs.mkShell {
+  buildInputs = [ pkgs.nodejs ];
+  shellHook = ''
+    export NODE_PATH=$(npm config get prefix)/lib/node_modules
+  '';
+}
+```
+
+In a real build derivation, you'd pin the exact Node version and even vendor
+`node_modules` to guarantee purity. That's not just possible---it's the default
+behavior. The build is a pure function. Run it anywhere with the same inputs,
+get the same output.
+
+Nix expressions are declarative. You describe what you want, not how to get
+there. Every package build is a function of its inputs: the source code, the
+compiler, the dependencies, the environment variables, and the build scripts.
+Nothing more. There's no hidden context. No reliance on the global state of your
+machine. No network access during builds unless you explicitly allow it, e.g.,
+with `--no-sandbox`.
+
+## Filesystem and Image Composition
+
+Docker builds up layers. Each command creates a filesystem diff. These are
+cached and stacked. But unless you're extremely careful, you accumulate garbage.
+Consider a typical pattern:
+
+```dockerfile
+RUN apt-get update && apt-get install -y build-essential python3 \
+    && rm -rf /var/lib/apt/lists/*
+```
+
+This trick is needed to keep images small. If you forget to clean up, that layer
+includes every downloaded `.deb` file and package index. There are also
+situations where `/var/lib/apt/lists/` _needs_ to remain in place.
+
+With Nix, you never need these hacks. The build output contains only what was
+declared. For example, here's a `default.nix` to build a Python application:
+
+```nix
+# XXX: The `import <nixpkgs>` pattern is actually discouraged, and it is usually
```

Comment:
Maybe skip all this and do a `callPackage`-able expr instead?

Reply:
`callPackage`-able as in? A more correct example is in the second post, I want this to be a quick and dirty example.